File: dart/hashmap/group_anagrams.dart
///Note: do not return sorted/re-arranged strs

class Solution1 {
  List<List<String>> groupAnagrams(List<String> strs) {
    final hashMap = <String, List<String>>{};
    strs.forEach((element) {
      final listStr = element.split('');
      listStr.sort();
      final key = listStr.join();
      if (hashMap.containsKey(key)) {
        hashMap[key]!.add(element);
      } else {
        hashMap[key] = [element];
      }
    });
    return hashMap.values.toList();
  }
}

class Solution2 {
  List<List<String>> groupAnagrams(List<String> strs) {
    final arr = <String>[];
    final hashMap = <String, List<String>>{};
    for (var i = 0; i < strs.length; i++) {
      final s = strs[i].split('');
      s.sort();
      arr.add(s.join());
    }

    for (var i = 0; i < arr.length; i++) {
      if (hashMap[arr[i]] == null) {
        hashMap[arr[i]] = [];
      }
      hashMap[arr[i]]?.add(strs[i]);
    }
    return hashMap.values.toList();
  }
}

void main() {
  final strs = <String>[
    "eat",
    "tea",
    "tan",
    "ate",
    "nat",
    "bat",
    "ba",
    "ca",
    "a",
    "ab",
  ];

  final sol = Solution2();
  sol.groupAnagrams(strs);
}

File: go/hashmap/group_anagrams.go
package hashmap

import (
	"fmt"
	"sort"
	"strings"
)

/// Note: do not return sorted/re-arranged strs

func getMapValues(m map[string][]string) [][]string {
	values := make([][]string, 0, len(m))
	for _, v := range m {
		values = append(values, v)
	}
	return values
}

func GroupAnagrams(strs []string) [][]string {
	hashMap := make(map[string][]string)
	for _, element := range strs {
		listStr := strings.Split(element, "")
		sort.Strings(listStr)
		key := strings.Join(listStr, "")
		hashMap[key] = append(hashMap[key], element)
	}
	res := getMapValues(hashMap)
	return res
}

type Key [26]byte

func strKey(str string) (key Key) {
	for i := range str {
		key[str[i]-'a']++
	}
	return key
}
func GroupAnagrams21(strs []string) [][]string {
	groups := make(map[Key][]string)
	for _, v := range strs {
		key := strKey(v)
		fmt.Println(key)
		groups[key] = append(groups[key], v)
	}
	result := make([][]string, 0, len(groups))
	for _, v := range groups {
		result = append(result, v)
	}
	return result
}
